### ISSUE TITLE
aliasing seemingly caused by prgba_to_bgra method??

### ISSUE THREAD

**Issue by marc2k3** [reupen/svg-services#5]
You might not notice/need to change it for your own use case but I did notice this...

![image](https://user-images.githubusercontent.com/62050527/204073487-d3c02366-1f7d-497c-aa43-18080d374aa8.png)

I found that removing it entirely and supplying `GUID_WICPixelFormat32bppPRGBA` to `CreateBitmapFromMemory` gives a better result...

![image](https://user-images.githubusercontent.com/62050527/204073591-29297df8-011a-447d-aaf0-754aae2429be.png)

**Comment by reupen** [comment]
I rendered the SVG from [this page](https://source.android.com/docs/setup/about/brands) in a giant Columns UI button and it looks OK to me:

![image](https://user-images.githubusercontent.com/12693549/204082912-6c9d4e9d-4254-4a16-a18c-135e7310c76d.png)

How are you rendering the image? Are you sure it doesn't require pre-multiplied alpha? (Just using `GUID_WICPixelFormat32bppPRGBA` won't cause a conversion.)

**Comment by marc2k3** [comment]
My first screenshot was using your original code untouched. Right colour but aliased.

The 2nd screenshot was removing the call to `prgba_to_bgra` but replacing the original `GUID_WICPixelFormat32bppPBGRA` with the `GUID_WICPixelFormat32bppPRGBA` which I thought "fixed" it but I really am trialling/erroring without really knowing what I'm doing. 😬 

By leaving the original `GUID_WICPixelFormat32bppPBGRA`, I get the same colour as your screenshot which I consider to be wrong??

This is the original in a browser and what I consider to be the "right" shade of green.

https://raw.githubusercontent.com/marc2k3/jscript-panel/0a510d3521a8ea0e78a642a4558947f785cb8f34/component/samples/svg/android.svg

This is my entire SVG reading method (basically your code with a few tweaks) ...

```cpp
static IJSImage* svg_to_image(wil::zstring_view str, uint32_t max_width)
{
	resvg_options* opt = resvg_options_create();
	resvg_render_tree* tree{};
	const int status = resvg_parse_tree_from_data(str.data(), str.size(), opt, &tree);
	resvg_options_destroy(opt);

	if (status == RESVG_OK)
	{
		resvg_fit_to fit_to{};
		const resvg_size size = resvg_get_image_size(tree);
		uint32_t width{}, height{};

		if (max_width == 0)
		{
			width = to_uint(size.width);
			height = to_uint(size.height);
			fit_to = resvg_fit_to(RESVG_FIT_TO_TYPE_ORIGINAL, 1.f);
		}
		else
		{
			width = max_width;
			height = to_uint(size.height / size.width * max_width);
			fit_to.type = RESVG_FIT_TO_TYPE_WIDTH;
			fit_to.value = to_float(width);
		}

		const uint32_t stride = width * 4U;
		std::vector<uint8_t> data(stride * height);
		wil::com_ptr_t<IWICBitmap> bitmap;

		resvg_render(tree, fit_to, resvg_transform_identity(), width, height, reinterpret_cast<char*>(data.data()));

		if (SUCCEEDED(g_imaging_factory->CreateBitmapFromMemory(width, height, GUID_WICPixelFormat32bppPRGBA, stride, stride * height, data.data(), &bitmap)))
		{
			return new ComObjectImpl<JSImage>(bitmap);
		}
	}
	return nullptr;
}
```

This is my DrawImage code from my yucky javascript-y component using, no gdi in sight. :P

```cpp
STDMETHODIMP JSGraphics::DrawImage(IJSImage* image, float dstX, float dstY, float dstW, float dstH, float srcX, float srcY, float srcW, float srcH, float opacity, float angle)
{
	RETURN_HR_IF(E_POINTER, !m_render_target);
	if (srcW < 1.f || srcH < 1.f) return S_OK;

	IWICBitmap* wic_bitmap = nullptr;
	RETURN_IF_FAILED(image->get(arg_helper(&wic_bitmap)));

	wil::com_ptr_t<ID2D1Bitmap> d2d_bitmap;
	wil::com_ptr_t<ID2D1DeviceContext> context;

	// wic bitmaps can seemingly be valid but fail here, returning S_OK to suppress errors :/
	if (FAILED(m_render_target->CreateBitmapFromWicBitmap(wic_bitmap, &d2d_bitmap))) return S_OK;
	RETURN_IF_FAILED(m_render_target->QueryInterface(IID_PPV_ARGS(&context)));

	if (angle == 0.f)
	{
		context->DrawBitmap(d2d_bitmap.get(), to_rectf(dstX, dstY, dstW, dstH), opacity, D2D1_INTERPOLATION_MODE_HIGH_QUALITY_CUBIC, to_rectf(srcX, srcY, srcW, srcH));
	}
	else
	{
		const D2D1_POINT_2F centre = { dstX + (dstW / 2.f), dstY + (dstH / 2.f) };
		context->SetTransform(D2D1::Matrix3x2F::Rotation(angle, centre));
		context->DrawBitmap(d2d_bitmap.get(), to_rectf(dstX, dstY, dstW, dstH), opacity, D2D1_INTERPOLATION_MODE_HIGH_QUALITY_CUBIC, to_rectf(srcX, srcY, srcW, srcH));
		context->SetTransform(D2D1::Matrix3x2F::Rotation(0.f, centre));
	}
	return S_OK;
}
```

**Comment by reupen** [comment]
The SVG I linked is actually a different shade of green to yours. Now the screenshot I took does also show in the wrong shade of green in the browser for me, but I'm guessing that's to do with colour profiles because the real thing matches the colour of the original (ignoring some some off-by-one errors).

At the moment the component is outputting `GUID_WICPixelFormat32bppBGRA` (no P), with GDI use in mind. So if you're going to switch back to this, it sounds like you would need/want an option in the API to output pre-multiplied alpha (GUID_WICPixelFormat32bppPRGBA) to match your D2D render target.
